db/seeds.d/16-mail_notifications.rb

```diff
@@ -60,8 +59,14 @@
     mail.attributes = notification
     mail.save! if mail.changed?
   else
-    created_notification = MailNotification.create(notification)
+    created_notification = notification[:name] == 'puppet_error_state' ? PuppetError.create(notification) : MailNotification.create(notification)
```

Comment:
Move this logic to `def initialize` (override it) on MailNotification. Depending on the name you can create a different type of MailNotification subclass (like Puppet error).